sos/report/plugins/gcp.py

```diff
@@ -0,0 +1,111 @@
+import json
+from http.client import HTTPResponse
+from typing import Union
+from urllib import request
+from urllib.error import URLError, HTTPError
+
+from sos.report.plugins import Plugin, IndependentPlugin
+from sos.report.plugins import SoSPredicate
+
+
+class GCP(Plugin, IndependentPlugin):
+
+    short_desc = 'Google Cloud Platform'
+
+    plugin_name = 'gcp'
+    profiles = ('virt',)
+
+    METADATA_ROOT = "http://metadata.google.internal/computeMetadata/v1/"
+
+    # A line we will be looking for in the dmesg output. If it's there,
+    # that means we're running on a Google Cloud Compute instance.
+    GOOGLE_DMI = "DMI: Google Google Compute Engine/Google " \
+                 "Compute Engine, BIOS Google"
+
+    def check_enabled(self):
+        """
+        Checks if this plugin should be executed at all. In this case, it
+        will check the `dmesg` command output to see if the system is
+        running on a Google Cloud Compute instance.
+        """
+        dmesg = self.exec_cmd("dmesg")
+        if dmesg['status'] != 0:
+            return False
+        return self.GOOGLE_DMI in dmesg['output']
+
+    def setup(self):
+        """
+        Collect the following info:
+         * Metadata from the Metadata server
+         * `gcloud auth list` output
+         * Any google services output from journal
+        """
+
+        # Capture gcloud auth list
+        predicate = SoSPredicate(self, packages=['google-cloud-sdk'])
+        self.add_cmd_output("gcloud auth list", pred=predicate, tags=['gcp'])
+
+        # Get and store Metadata
+        try:
+            metadata = json.dumps(self.get_metadata(), indent=4)
+            self.add_string_as_file(metadata, "metadata.json", plug_dir=True,
+                                    tags=['gcp'])
+        except RuntimeError as err:
+            self.add_string_as_file(str(err), 'metadata.json',
+                                    plug_dir=True, tags=['gcp'])
+
+        # Add journal entries
+        self.add_journal(units="google*", tags=['gcp'])
+
+    def get_metadata(self) -> dict:
+        try:
+            response = self._query_address(self.METADATA_ROOT)
+            if response.code != 200:
+                raise RuntimeError(
+                    f"Failed to communicate with Metadata Server "
+                    f"(code: {response.code}): " + response.read().decode())
+        except URLError as err:
+            raise RuntimeError("Failed to communicate with Metadata Server: "
+                               + str(err))
+
+        return self._get_metadata(self.METADATA_ROOT)
+
+    @staticmethod
+    def _query_address(url: str) -> HTTPResponse:
+        """
+        Query the given url address with headers required by Google Metadata
+        Server.
+        """
+        req = request.Request(url, headers={'Metadata-Flavor': 'Google'})
+        return request.urlopen(req)
+
+    @classmethod
+    def _get_metadata(cls, url: str = '') -> Union[dict, str]:
```

Comment:
Putting my paranoic hat on: we call some URI in a directory-recursive way. What if - due to a proxy or network or DNS issue - each query takes many seconds, causing the whole `self._get_metadata(self.METADATA_ROOT)` takes minutes?

(but I have no idea how to easily protect that behaviour to happen, so I think we can live with this tiny risk)

Reply:
The thing about querying Metadata Server on GCP is that the requests **never** leave the physical machine. The metadata server is just a simple process running next to the VM on the physical host. It'd be an extremely weird event for those local requests to be slow.

DNS is also not an issue, since `metadata.google.internal` name gets resolved locally. 

Because the metadata server is so close to the VM, it can be used to acquire authentication token for GCP Service Accounts (that's why I redact them in line 91). This method will end fast, either with an error or success.